File: python/002_Add_Two_Numbers.py
# You are given two linked lists representing two non-negative numbers. 
# The digits are stored in reverse order and each of their nodes contain a single digit.
# Add the two numbers and return it as a linked list.

# Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
# Output: 7 -> 0 -> 8

# Subscribe to see which companies asked this question

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Runtime: 125 ms

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        num1, num2 = l1, l2
        carry = 0
        ans = ListNode("dummy")
        this_node = ans
        while num1 or num2 or carry:
            this_sum = 0
            if num1:
                this_sum += num1.val
                num1 = num1.next
            if num2:
                this_sum += num2.val
                num2 = num2.next
            if carry:
                this_sum += carry
            this_node.next = ListNode(this_sum%10)
            carry = this_sum >= 10
            this_node = this_node.next

        return ans.next

File: java/002_Add_Two_Numbers.java
// You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

// You may assume the two numbers do not contain any leading zero, except the number 0 itself.

// Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
// Output: 7 -> 0 -> 8
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummyNode = new ListNode(-1);
        int carry = 0;
        ListNode tmp1 = l1, tmp2 = l2, tmpAns = dummyNode;
        while (tmp1!= null || tmp2 != null || carry!=0){
            int thisDigit = carry;
            if (tmp1 != null){
                thisDigit += tmp1.val;
                tmp1 = tmp1.next;
            }
            if (tmp2 != null){
                thisDigit += tmp2.val;
                tmp2 = tmp2.next;
            }
            carry = thisDigit / 10;
            thisDigit = thisDigit % 10;
            tmpAns.next = new ListNode(thisDigit);
            tmpAns = tmpAns.next;
        }
        return dummyNode.next;
        
    }
}